mlir/lib/Dialect/QIR/Builder/QIRProgramBuilder.cpp

```diff
@@ -0,0 +1,460 @@
+        builder, loc, static_cast<int64_t>(metadata_.numResults));
+    // Cache for reuse
+    resultPointerCache[metadata_.numResults] = resultValue;
+    registerResultMap.insert({key, resultValue});
+    metadata_.numResults++;
+  }
+
+  // Create mz call
+  const auto mzSignature = LLVM::LLVMFunctionType::get(
+      LLVM::LLVMVoidType::get(builder.getContext()), {ptrType, ptrType});
+  auto mzDecl =
+      getOrCreateFunctionDeclaration(builder, module, QIR_MEASURE, mzSignature);
+  builder.create<LLVM::CallOp>(loc, mzDecl, ValueRange{qubit, resultValue});
+
+  return *this;
+}
+
+QIRProgramBuilder& QIRProgramBuilder::reset(const Value qubit) {
+  // Save current insertion point
+  const OpBuilder::InsertionGuard insertGuard(builder);
+
+  // Insert in measurements block (before branch)
+  builder.setInsertionPoint(measurementsBlock->getTerminator());
+
+  // Create reset call
+  const auto qirSignature = LLVM::LLVMFunctionType::get(
+      LLVM::LLVMVoidType::get(builder.getContext()),
+      LLVM::LLVMPointerType::get(builder.getContext()));
+  auto fnDecl =
+      getOrCreateFunctionDeclaration(builder, module, QIR_RESET, qirSignature);
+  builder.create<LLVM::CallOp>(loc, fnDecl, ValueRange{qubit});
+
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+// Unitary Operations
+//===----------------------------------------------------------------------===//
+
+QIRProgramBuilder& QIRProgramBuilder::x(const Value qubit) {
+  // Save current insertion point
+  const OpBuilder::InsertionGuard insertGuard(builder);
+
+  // Insert in body block (before branch)
+  builder.setInsertionPoint(bodyBlock->getTerminator());
+
+  // Create x call
+  const auto qirSignature = LLVM::LLVMFunctionType::get(
+      LLVM::LLVMVoidType::get(builder.getContext()),
+      LLVM::LLVMPointerType::get(builder.getContext()));
+  auto fnDecl =
+      getOrCreateFunctionDeclaration(builder, module, QIR_X, qirSignature);
+  builder.create<LLVM::CallOp>(loc, fnDecl, ValueRange{qubit});
+
+  return *this;
+}
+
+QIRProgramBuilder& QIRProgramBuilder::rx(Value theta, const Value qubit) {
+  // Save current insertion point
+  const OpBuilder::InsertionGuard insertGuard(builder);
+
+  // Insert in body block (before branch)
+  builder.setInsertionPoint(bodyBlock->getTerminator());
+
+  // Create rx call
+  const auto qirSignature = LLVM::LLVMFunctionType::get(
+      LLVM::LLVMVoidType::get(builder.getContext()),
+      {LLVM::LLVMPointerType::get(builder.getContext()),
+       Float64Type::get(builder.getContext())});
+  auto fnDecl =
+      getOrCreateFunctionDeclaration(builder, module, QIR_RX, qirSignature);
+  builder.create<LLVM::CallOp>(loc, fnDecl, ValueRange{qubit, theta});
+
+  return *this;
+}
+QIRProgramBuilder& QIRProgramBuilder::rx(double theta, const Value qubit) {
+  // Create constant for theta
+  auto thetaConst =
+      builder.create<LLVM::ConstantOp>(loc, builder.getF64FloatAttr(theta));
```

Comment:
This instant might have to be created in the top-level block for the tests to properly work.